daliuge-engine/dlg/manager/proc_daemon.py

```diff
@@ -201,58 +208,80 @@ def startDIM(self, nodes):
+            )
         return
 
     def startMM(self):
         tool = get_tool()
         args = ["--host", "0.0.0.0"]
         args += self._verbosity_as_cmdline()
-        logger.info("Starting Master Drop Manager with args: %s" % (" ".join(args)))
+        logger.info("Starting Master Drop Manager with args: %s", (" ".join(args)))
         self._mm_proc = tool.start_process("mm", args)
-        logger.info("Started Master Drop Manager with PID %d" % (self._mm_proc.pid))
+        logger.info("Started Master Drop Manager with PID %d", self._mm_proc.pid)
 
         # Also subscribe to zeroconf events coming from NodeManagers and feed
         # the Master Manager with the new hosts we find
         if self._zeroconf:
-            mm_client = client.MasterManagerClient()
-            node_managers = {}
+            nm_assigner = NMAssigner()
 
-            def nm_callback(zeroconf, service_type, name, state_change):
+            def _callback(zeroconf, service_type, name, state_change, adder, remover, accessor):
                 info = zeroconf.get_service_info(service_type, name)
                 if state_change is zc.ServiceStateChange.Added:
                     server = socket.inet_ntoa(_get_address(info))
                     port = info.port
-                    node_managers[name] = (server, port)
+                    adder(name, server, port)
                     logger.info(
-                        "Found a new Node Manager on %s:%d, will add it to the MM"
-                        % (server, port)
+                        "Found a new %s on %s:%d, will add it to the MM",
+                        service_type, server, port
                     )
-                    mm_client.add_node(server)
                 elif state_change is zc.ServiceStateChange.Removed:
-                    server, port = node_managers[name]
+                    server, port = accessor(name)
                     logger.info(
-                        "Node Manager on %s:%d disappeared, removing it from the MM"
-                        % (server, port)
+                        "%s on %s:%d disappeared, removing it from the MM",
+                        service_type, server, port
                     )
 
                     # Don't bother to remove it if we're shutting down. This way
                     # we avoid hanging in here if the MM is down already but
                     # we are trying to remove our NM who has just disappeared
                     if not self._shutting_down:
-                        try:
-                            mm_client.remove_node(server)
-                        finally:
-                            del node_managers[name]
+                        remover(name)
+
+            nm_callback = functools.partial(_callback, service_type='NodeManager', adder=nm_assigner.add_nm,
+                                            remover=nm_assigner.remove_nm,
+                                            accessor=nm_assigner.get_nm)
```

Comment:
Sorry, I wasn't very clear on my previous comment: when I said that `service_type` "should have" `NodeManager` or `DIM` I meant that those values will already be received by the callback when invoked by `zeroconf`, not that you had to provide them explicitly here. In fact when you create the service browser below, the second argument is the `service_type` that will eventually be passed to the callback. The zeroconf documentation is not explicit about this, but there are [examples](https://github.com/jstasiak/python-zeroconf/blob/1aa7842ae0f914c10465ae977551698046406d55/examples/browser.py#L68
) that show this is how it works.

As currently written this works only by chance: since the `service_type` you are passing here has the exact same value than the corresponding one you give in `browse_service` there's no error, but anything different would raise an exception.

In other words: let's avoid the confusion, and not pass `service_type='NodeManager'` here, same for the DIM callback below.